src/linux/write_zeroes.rs

```diff
@@ -32,150 +35,276 @@ impl PunchHole for File {
+        // fallocate() failed; fall back to writing a buffer of zeroes until we have written up
+        // to `length`.
         let buf_size = min(length, 0x10000);
         let buf = vec![0u8; buf_size];
-        let mut nwritten: usize = 0;
-        while nwritten < length {
-            let remaining = length - nwritten;
+        let mut num_written: usize = 0;
+        while num_written < length {
+            let remaining = length - num_written;
             let write_size = min(remaining, buf_size);
-            nwritten += self.write(&buf[0..write_size])?;
+            num_written += self.write_at(&buf[0..write_size], offset + num_written as u64)?;
         }
         Ok(length)
     }
 }
 
+impl<T: WriteZeroesAt + Seek> WriteZeroes for T {
+    fn write_zeroes(&mut self, length: usize) -> Result<usize> {
+        let offset = self.seek(SeekFrom::Current(0))?;
+        let num_written = self.write_zeroes_at(offset, length)?;
+        // Advance the seek cursor as if we had done a real write().
+        self.seek(SeekFrom::Current(num_written as i64))?;
+        Ok(length)
+    }
+}
+
 #[cfg(test)]
-#[allow(clippy::unused_io_amount)]
 mod tests {
     use super::*;
-    use std::fs::OpenOptions;
-    use std::io::{Read, Seek, SeekFrom};
-    use std::path::PathBuf;
 
-    use crate::tempdir::TempDir;
+    use std::io::{Read, Seek, SeekFrom, Write};
+
+    use crate::tempfile::TempFile;
 
     #[test]
-    fn simple_test() {
-        let tempdir = TempDir::new_with_prefix("/tmp/write_zeroes_test").unwrap();
-        let mut path = PathBuf::from(tempdir.as_path());
-        path.push("file");
-        let mut f = OpenOptions::new()
-            .read(true)
-            .write(true)
-            .create(true)
-            .open(&path)
-            .unwrap();
+    fn test_small_write_zeroes() {
+        const NON_ZERO_VALUE: u8 = 0x55;
+
+        let mut f = TempFile::new().unwrap().into_file();
         f.set_len(16384).unwrap();
 
-        // Write buffer of non-zero bytes to offset 1234
-        let orig_data = [0x55u8; 5678];
+        // Write buffer of non-zero bytes to offset 1234.
+        let orig_data = [NON_ZERO_VALUE; 5678];
         f.seek(SeekFrom::Start(1234)).unwrap();
         f.write(&orig_data).unwrap();
 
-        // Read back the data plus some overlap on each side
+        // Read back the data plus some overlap on each side.
         let mut readback = [0u8; 16384];
         f.seek(SeekFrom::Start(0)).unwrap();
         f.read(&mut readback).unwrap();
-        // Bytes before the write should still be 0
-        for read in readback[0..1234].iter() {
+        // Bytes before the write should still be 0.
+        for read in &readback[0..1234] {
             assert_eq!(*read, 0);
         }
-        // Bytes that were just written should be 0x55
-        for read in readback[1234..(1234 + 5678)].iter() {
-            assert_eq!(*read, 0x55);
+        // Bytes that were just written should be 0x55.
```

Comment:
nit: we are not using a hardcoded value, so we need to update the comment as well.

Reply:
I left the numeric value here since it sounded more decent :)). Updated the comment.